File: backend/app/schemas/reminder.py
"""Pydantic schemas for Vehicle Reminder operations."""

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class ReminderCreate(BaseModel):
    """Schema for creating a vehicle reminder."""

    title: str = Field(..., min_length=1, max_length=200)
    reminder_type: Literal["date", "mileage", "both", "smart"]
    due_date: date | None = None
    due_mileage: int | None = Field(None, gt=0)
    notes: str | None = None
    line_item_id: int | None = None

    @model_validator(mode="after")
    def validate_fields_for_type(self) -> "ReminderCreate":
        """Ensure required fields are present based on reminder type."""
        if self.reminder_type in ("date", "both", "smart") and not self.due_date:
            raise ValueError("due_date required for this reminder type")
        if self.reminder_type in ("mileage", "both", "smart") and not self.due_mileage:
            raise ValueError("due_mileage required for this reminder type")
        return self


class ReminderUpdate(BaseModel):
    """Schema for updating a vehicle reminder.

    Status is NOT here — use /done or /dismiss endpoints.
    Validation is lenient (fields may be absent). The route handler merges
    this patch onto the existing reminder and validates the final state.
    """

    title: str | None = Field(None, min_length=1, max_length=200)
    reminder_type: Literal["date", "mileage", "both", "smart"] | None = None
    due_date: date | None = None
    due_mileage: int | None = Field(None, gt=0)
    notes: str | None = None


class ReminderResponse(BaseModel):
    """Schema for reminder response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    vin: str
    line_item_id: int | None
    title: str
    reminder_type: str
    due_date: date | None
    due_mileage: int | None
    status: str
    notes: str | None
    estimated_due_date: date | None = None
    last_notified_at: datetime | None
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/reminder.ts
/**
 * Vehicle Reminder type definitions
 */

export type ReminderType = 'date' | 'mileage' | 'both' | 'smart'
export type ReminderStatus = 'pending' | 'done' | 'dismissed'

export interface Reminder {
  id: number
  vin: string
  line_item_id: number | null
  title: string
  reminder_type: ReminderType
  due_date: string | null
  due_mileage: number | null
  status: ReminderStatus
  notes: string | null
  estimated_due_date: string | null
  last_notified_at: string | null
  created_at: string
  updated_at: string
}

export interface ReminderCreate {
  title: string
  reminder_type: ReminderType
  due_date?: string
  due_mileage?: number
  notes?: string
  line_item_id?: number
}

export interface ReminderUpdate {
  title?: string
  reminder_type?: ReminderType
  due_date?: string
  due_mileage?: number
  notes?: string
}

export interface ReminderDraft extends ReminderCreate {
  enabled: boolean
}
